### DISCUSSION TITLE
Recommended Snakemake practices

### DISCUSSION THREAD

**Discussion by AndyMcAliley** [USGS-R/snakemake-tutorial-scratch#12] [Ideas]
Here's a place to compile a self-consistent set of practices for using Snakemake. The idea is to have these practices form a loose style guide. It's not the best or the only way to use Snakemake, but it's a way to use Snakemake that has been successful for USGS data scientists. Someone who is new to Snakemake can apply these practices to their data science project and expect a reasonable fit. We hope this saves newcomers to Snakemake the hassle of sorting through several usage paradigms and trying to pick one that suits their use case, while simultaneously trying to learn the ropes of Snakemake.

**Comment by AndyMcAliley** [comment]
Snakemake is file-based: inputs and outputs for each rule are usually files. Unlike some other pipeline tools, snakemake does not allow code objects to be inputs or outputs. So, when planning out a pipeline, it makes sense to think first and foremost in terms of what files will be generated by a rule, and what files will be needed for that rule to run.

**Comment by AndyMcAliley** [comment]
If multiple files are produced by a snakemake rule, be sure to list all files as outputs of that rule. Snakemake may raise an error if it detects new files have been created that it didn't expect - especially if those files are used as inputs to downstream rules.

**Comment by AndyMcAliley** [comment]
There are a few different directives that can be used for a rule to execute code: `shell`, `script`, and `run`. 

- `shell` is best for actions that don't require any custom code, only shell commands (copying files, moving files, external programs, etc.)
- `script` is best for running custom Python/R code. If the script is changed, snakemake will detect the change and issue a warning.
- `run` can be useful for interacting with a library, especially when each snakemake rule can be completed by executing one function in that library. However, unlike the `script` directive, if changes are made to the code in the library, snakemake won't detect the change. Therefore, we recommend favoring `script` over `run`.

**Comment by AndyMcAliley** [comment]
Inputs and outputs to rules can be named. We recommend naming all inputs and outputs whenever practical, like so:
```
input:
    in_file = "path/to/input.csv"
output:
    out_file = "path/to/output.csv"
```

**Comment by AndyMcAliley** [comment]
Variables get passed among many places and scopes: the Snakefile, a Python script, inside a `main()` function, and inside other functions. It helps to keep the variable names consistent throughout.

**Comment by ted80810** [comment]
When there are multiple inputs in a rule using "expand" with "wildcards" is particularly useful to capture every file.
ie. 
```
timesteps = ["0h", "3h", "6h", "9h", "12h", "15h", "18h", "21h", "24h"]
ensemble = [1, 2, 3,4,5]

rule hourly_temp:
    input: expand("{timepoints}_{ensembles}_temp.csv", timesteps=timesteps, ensembles=ensembles)
```

Above, we build the file name with wildcards and iterate over every file with expand.
